internal/webhook/external/azure.go

```diff
@@ -0,0 +1,188 @@
+
+// getSecretValues implements WebhookReceiver.
+func (q *azureWebhookReceiver) getSecretValues(
+	secretData map[string][]byte,
+) ([]string, error) {
+	secretValue, ok := secretData[azureSecretDataKey]
+	if !ok {
+		return nil,
+			errors.New("Secret data is not valid for an Azure WebhookReceiver")
+	}
+	return []string{string(secretValue)}, nil
+}
+
+// getHandler implements WebhookReceiver.
+func (q *azureWebhookReceiver) getHandler(requestBody []byte) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		logger := logging.LoggerFromContext(ctx)
+		logger.Debug("received azure event", "userAgent", r.UserAgent())
+
+		switch {
+		// format is AzureContainerRegistry/<version>
+		case strings.Contains(r.UserAgent(), "AzureContainerRegistry"):
+			var event acrEvent
+			if err := json.Unmarshal(requestBody, &event); err != nil {
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(errors.New("invalid request body"),
+						http.StatusBadRequest,
+					),
+				)
+				return
+			}
+
+			switch event.Action {
+			case "push":
+				repoURL := fmt.Sprintf("%s/%s",
+					event.Request.Host,
+					event.Target.Repository,
+				)
+				if event.Target.MediaType == imageMediaType {
+					repoURL = image.NormalizeURL(repoURL)
+				} else {
+					repoURL = helm.NormalizeChartRepositoryURL(repoURL)
+				}
+				logger = logger.WithValues("repoURL", repoURL)
+				ctx = logging.ContextWithLogger(ctx, logger)
+				refreshWarehouses(ctx, w, q.client, q.project, repoURL)
+			case "ping":
+				xhttp.WriteResponseJSON(
+					w,
+					http.StatusOK,
+					map[string]string{
+						"msg": "ping event received, webhook is configured correctly",
+					},
+				)
+			default:
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(
+						fmt.Errorf("event type %s is not supported", event.Action),
+						http.StatusNotImplemented,
+					),
+				)
+				return
+			}
+		// Format is VSServices/<version>
+		case strings.Contains(r.UserAgent(), "VSServices"):
+			var event azureDevOpsEvent
+			if err := json.Unmarshal(requestBody, &event); err != nil {
+				xhttp.WriteErrorJSON(
+					w,
+					xhttp.Error(errors.New("invalid request body"),
+						http.StatusBadRequest,
+					),
+				)
+				return
+			}
+			repoURL := git.NormalizeURL(event.Resource.Repository.RemoteURL)
+			logger = logger.WithValues("repoURL", repoURL)
+			ctx = logging.ContextWithLogger(ctx, logger)
+			refreshWarehouses(ctx, w, q.client, q.project, repoURL)
+		default:
+			xhttp.WriteErrorJSON(
+				w,
+				xhttp.Error(
+					fmt.Errorf("user agent %s is not supported", r.UserAgent()),
+					http.StatusNotImplemented,
+				),
+			)
+			return
+		}
+	})
+}
+
+// For more information on the payload schema for Azure Container Registry, see:
+//
+//	Azure Container Registry
+//		https://learn.microsoft.com/en-us/azure/container-registry/container-registry-webhook-reference#payload-example-image-push-event
+//
+// nolint:lll
+type acrEvent struct {
```

Comment:
By widely observed convention, this comment should start with the name of the type.